### ISSUE TITLE
Possibile miglioramento

### ISSUE THREAD

**Issue by falci11** [kyleskom/NBA-Machine-Learning-Sports-Betting#501]
Il codice mescola casualmente (train_test_split) partite di anni diversi per formare i set di training e test. Questo fa sì che il modello possa "vedere" partite future durante l'addestramento, falsando i risultati. È il problema più grave. | x_train, x_test... = train_test_split(data, ..., test_size=.1)(Tutti i file, nessun parametro shuffle=False o ordinamento per data).
Ha una valutazione Ingannevole perchè La performance dichiarata (~69%) deriva da questo test set "contaminato" e non riflette la capacità di prevedere partite future. In un backtest temporale corretto, l'accuratezza potrebbe essere molto più bassa (es. 55-60%). | Ciclo che salva il modello solo se batte il record di accuratezza (max(acc_results)), selezionando il picco casuale su una suddivisione favorevole.
Ha Iperparametri Potenzialmente Non Ottimali infatti gli iperparametri sono fissati a valori arbitrari senza un processo di ottimizzazione strutturato su un periodo di validazione temporale. I max_depth': 20 (XGBoost_UO) è molto alto e rischia overfitting. Nessuna ricerca tramite GridSearchCV o RandomizedSearchCV. Non ha nessuna Gestione dello Sbilanciamento di Classe quindi il dataset potrebbe non avere un bilanciamento uniforme tra vittorie in casa e in trasferta. Alcuni modelli (come la regressione logistica) potrebbero favorire la classe maggioritaria senza un trattamento specifico. Non c'è nessuna analisi della distribuzione della variabile target (margin o OU) e nessun uso di parametri come class_weight.
Il codice attuale funziona, ma i suoi risultati sono tecnicamente non validi per il mondo reale a causa della violazione del principio fondamentale di non usare dati futuri. La priorità assoluta è implementare una suddivisione temporale rigorosa e rivalutare le performance.
Se riscriviamo Create_Games.py:
import os
import sqlite3
import sys
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import toml

sys.path.insert(1, os.path.join(sys.path[0], '../..'))
from src.Utils.Dictionaries import team_index_07, team_index_08, team_index_12, team_index_13, team_index_14, \
    team_index_current

config = toml.load("../../config.toml")

# 1. INIZIALIZZAZIONE CON LOGICA TEMPORALE
# ----------------------------------------
# Utilizziamo liste separate per tracciare ogni dato in modo ordinato.
dates_list = []          # Data della partita (FONDAMENTALE)
scores = []
win_margin = []
OU = []
OU_Cover = []
games = []               # Le feature delle squadre
days_rest_away = []
days_rest_home = []
home_team_names = []     # Nome della squadra di casa (per debug/tracciabilità)
away_team_names = []     # Nome della squadra in trasferta

teams_con = sqlite3.connect("../../Data/TeamData.sqlite")
odds_con = sqlite3.connect("../../Data/OddsData.sqlite")

for key, value in config['create-games'].items():
    print(f"Elaborazione stagione: {key}")
    odds_df = pd.read_sql_query(f"select * from \"odds_{key}_new\"", odds_con, index_col="index")
    
    # 2. ORDINAMENTO TEMPORALE CRITICO
    # ---------------------------------
    # Assicuriamo che le partite siano processate in ordine cronologico.
    # Convertiamo in datetime e ordiniamo. 'Date' è il nome della colonna in odds_df.
    if 'Date' not in odds_df.columns:
        # Se il formato del nome è diverso, adattare qui
        print(f"ERRORE: Colonna 'Date' non trovata in odds_{key}_new")
        continue
    odds_df['Date'] = pd.to_datetime(odds_df['Date'])
    odds_df = odds_df.sort_values('Date')
    
    team_table_str = key
    season = key

    for row in odds_df.itertuples():
        home_team = row.Home
        away_team = row.Away
        date = row.Date  # Questo ora è un oggetto datetime di Pandas/Timestamp
        
        # 3. REGISTRAZIONE DELLA DATA
        # ----------------------------
        # Salviamo la data PRIMA di qualsiasi altra operazione.
        dates_list.append(date)

        team_df = pd.read_sql_query(f"select * from \"{date.date()}\"", teams_con, index_col="index")
        if len(team_df.index) != 30:
            # Se non ci sono dati per tutte le 30 squadre in questa data, skip
            # Rimuoviamo l'ultima data aggiunta perché la partita non è valida
            dates_list.pop()
            continue

        # 4. REGISTRAZIONE DEI RISULTATI (TARGET)
        # ----------------------------------------
        scores.append(row.Points)
        OU.append(row.OU)
        days_rest_home.append(row.Days_Rest_Home)
        days_rest_away.append(row.Days_Rest_Away)
        home_team_names.append(home_team)
        away_team_names.append(away_team)

        # TARGET 1: Vittoria in Casa (1 = Vittoria, 0 = Sconfitta)
        win_margin.append(1 if row.Win_Margin > 0 else 0)

        # TARGET 2: Under/Over Cover (0 = Under, 1 = Over, 2 = Push)
        if row.Points < row.OU:
            OU_Cover.append(0)
        elif row.Points > row.OU:
            OU_Cover.append(1)
        else:
            OU_Cover.append(2)

        # 5. RECUPERO FEATURE DELLE SQUADRE (invariato dalla tua logica)
        # --------------------------------------------------------------
        if season in ['2019-20', '2020-21', '2021-22']:
            home_team_mapped = home_team
            away_team_mapped = away_team
            
            if home_team == 'Charlotte Hornets':
                home_team_mapped = 'Charlotte Bobcats'
            if away_team == 'Charlotte Hornets':
                away_team_mapped = 'Charlotte Bobcats'
                
            home_team_series = team_df.iloc[team_index_14.get(home_team_mapped)]
            away_team_series = team_df.iloc[team_index_14.get(away_team_mapped)]
            
        elif season in ['2022-23', '2023-24', '2024-25', '2025-26']:
            home_team_series = team_df.iloc[team_index_current.get(home_team)]
            away_team_series = team_df.iloc[team_index_current.get(away_team)]
        else:
            print(f"Stagione non supportata: {season}")
            continue

        game = pd.concat([home_team_series, away_team_series.rename(
            index={col: f"{col}.1" for col in team_df.columns.values}
        )])
        games.append(game)
            
odds_con.close()
teams_con.close()

# 6. COSTRUZIONE DEL DATASET FINALE TEMPORALMENTE ORDINATO
# --------------------------------------------------------
print("\nCostruzione del dataset finale...")
season_df = pd.concat(games, ignore_index=True, axis=1)
season_df = season_df.T

# Rimozione colonne non numeriche (assicurati che i nomi siano corretti)
columns_to_drop = ['TEAM_NAME', 'TEAM_NAME.1']
for col in ['TEAM_ID', 'TEAM_ID.1', 'Date', 'Date.1']:
    if col in season_df.columns:
        columns_to_drop.append(col)

frame = season_df.drop(columns=columns_to_drop)

# Aggiunta di tutte le colonne target e metadati
frame['Date'] = pd.to_datetime(dates_list)  # COLONNA TEMPORALE PRIMARIA
frame['Home_Team'] = home_team_names
frame['Away_Team'] = away_team_names
frame['Score'] = np.asarray(scores, dtype=float)
frame['Home-Team-Win'] = np.asarray(win_margin, dtype=float)
frame['OU'] = np.asarray(OU, dtype=float)
frame['OU-Cover'] = np.asarray(OU_Cover, dtype=float)
frame['Days-Rest-Home'] = np.asarray(days_rest_home, dtype=float)
frame['Days-Rest-Away'] = np.asarray(days_rest_away, dtype=float)

# 7. ORDINAMENTO FINALE E CREAZIONE INDICE TEMPORALE
# --------------------------------------------------
frame = frame.sort_values('Date').reset_index(drop=True)
frame['Game_ID'] = range(1, len(frame) + 1)  # ID progressivo basato sul tempo

# 8. CONVERSIONE TIPI PER PERFORMANCE
# ------------------------------------
for column in frame.columns:
    if frame[column].dtype == 'object':
        try:
            frame[column] = pd.to_numeric(frame[column])
        except:
            pass  # Tiene colonne stringa come i nomi delle squadre

# 9. SALVATAGGIO CON NOME CHIARO
# -------------------------------
con = sqlite3.connect("../../Data/dataset.sqlite")
frame.to_sql("dataset_temporal_correct", con, if_exists="replace")  # Nome esplicito
con.close()

print("✅ Dataset creato con successo!")
print(f"   • Partite totali: {len(frame)}")
print(f"   • Periodo: dal {frame['Date'].min().date()} al {frame['Date'].max().date()}")
print(f"   • Colonne: {len(frame.columns)}")
print(f"   • Salvato come: 'dataset_temporal_correct'")

Poi potremmo riscrivere i file di Training per farli allenare in modo temporale e non casuale perchè il problema fondamentale è lo stesso in tutti: usano train_test_split con mescolamento casuale, violando il principio temporale dei dati sportivi. Questo rende le loro valutazioni completamente inaffidabili per l'uso reale. Tu cosa ne pensi in merito a tutto ciò?
PS:Ho provato a riscriverli adattandoli a Create_Games.py aggiornato
